### PR TITLE
more robust pipeline feature extraction

### PR DESCRIPTION
**Pull request by tmchartrand**
Most changes are minor restructuring and error handling to allow partial feature extraction when part of the pipeline fails.
A few other minor error handling fixes are also included.

I did get a few test failures:
Two of the pipeline output tests will need to be updated (Npr3 and Sst), as those cells had fx or qc failures in the current reference, and no features extracted.
The third pipeline output failure (Pvalb) is an hdmf issue unrelated to these changes.

### CODE REVIEW

File: ipfx/bin/pipeline_from_specimen_id.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ def main():
 
     ju.write(os.path.join(cell_dir,OUTPUT_JSON), pipe_output)
 
-if __name__ == "__main__": main()
+if __name__ == "__main__": run_pipeline_from_id(sys.argv[1])
```
Comment (NileGraddis) [inline ipfx/bin/pipeline_from_specimen_id.py]: could you instead do:
```
import argparse
parser = argparse.ArgumentParser(description="do some stuff")
parser.add_argument("specimen_id", type=int, help="identifies specimen on which to run pipeline")
run_pipeline_from_id(args.specimen_id)
```

File: ipfx/bin/run_feature_extraction.py [hunk 2]
```diff
@@ -65,24 +65,16 @@ def run_feature_extraction(input_nwb_file,
...
-        cell_state = {"failed_fx": False, "fail_fx_message": None}
+    cell_features, sweep_features, cell_record, sweep_records, cell_state = dsft.extract_data_set_features(data_set)
 
-        feature_data = { 'cell_features': cell_features,
-                         'sweep_features': sweep_features,
-                         'cell_record': cell_record,
-                         'sweep_records': sweep_records,
-                         'cell_state': cell_state
-                         }
+    if cell_info: cell_record.update(cell_info)
```
Comment (NileGraddis) [inline ipfx/bin/run_feature_extraction.py]: I know this was a single line in the original, but could you put the `cell_record.update(cell_info)` on the next line?

File: ipfx/data_set_features.py [hunk 3]
```diff
@@ -162,66 +162,97 @@ def extract_cell_features(data_set,
...
-    ssq_sweeps.align_to_start_of_epoch("experiment")
-
-    ssq_start, ssq_dur, _, _, _ = stf.get_stim_characteristics(ssq_sweeps.sweeps[0].i, ssq_sweeps.sweeps[0].t)
-    ssq_spx, ssq_spfx = extractors_for_sweeps(ssq_sweeps,
-                                              est_window = [ssq_start,ssq_start+0.001],
-                                              **detection_parameters(data_set.SHORT_SQUARE))
-    ssq_an = spa.ShortSquareAnalysis(ssq_spx, ssq_spfx)
-    ssq_features = ssq_an.analyze(ssq_sweeps)
-    cell_features["short_squares"] = ssq_an.as_dict(ssq_features, [ dict(sweep_number=sn) for sn in short_square_sweep_numbers ])
+    try:
```
Comment (NileGraddis) [inline ipfx/data_set_features.py]: These all seem extremely similar, particularly the error handling code, which looks to be exactly duplicated up to a single string. I think these common components ought to be refactored into reusable functions.

File: ipfx/data_set_features.py [hunk 4]
```diff
@@ -162,66 +162,97 @@ def extract_cell_features(data_set,
...
-        raise er.FeatureError("Could not find hero sweep.")
+    try:
+        lu.log_pretty_header("Long Squares:",level=2)
+        if len(long_square_sweep_numbers) == 0:
+            raise er.FeatureError("No long_square sweeps available for feature extraction")
+
+        lsq_sweeps = data_set.sweep_set(long_square_sweep_numbers)
+        lsq_sweeps.select_epoch("recording")
+        lsq_sweeps.align_to_start_of_epoch("experiment")
+        # TODO: fail properly when stim sweeps differ in start/duration
```
Comment (NileGraddis) [inline ipfx/data_set_features.py]: is this TODO complicated to implement? If so, could you create an issue documenting what needs to happen?

File: ipfx/data_set_features.py [hunk 5]
```diff
@@ -162,66 +162,97 @@ def extract_cell_features(data_set,
...
+        if len(short_square_sweep_numbers) == 0:
+            raise er.FeatureError("No short square sweeps available for feature extraction")
+
+        ssq_sweeps = data_set.sweep_set(short_square_sweep_numbers)
+        ssq_sweeps.select_epoch("recording")
+        ssq_sweeps.align_to_start_of_epoch("experiment")
+
+        ssq_start, ssq_dur, _, _, _ = stf.get_stim_characteristics(ssq_sweeps.sweeps[0].i, ssq_sweeps.sweeps[0].t)
+        ssq_spx, ssq_spfx = extractors_for_sweeps(ssq_sweeps,
+                                                est_window = [ssq_start,ssq_start+0.001],
```
Comment (NileGraddis) [inline ipfx/data_set_features.py]: What is 0.001? Can it be a parameter of this function?

File: ipfx/data_set_features.py [hunk 6]
```diff
@@ -308,21 +339,23 @@ def extract_data_set_features(data_set, subthresh_min_amp=None):
...
+                                        lsq_sweep_numbers,
+                                        subthresh_min_amp)
+    # TODO: looks like this is adding peak_deflect, but could be refactored not to need cell_features
     # shuffle peak deflection for the subthreshold long squares
-    for s in cell_features["long_squares"]["subthreshold_sweeps"]:
-        sweep_features[s['sweep_number']]['peak_deflect'] = s['peak_deflect']
-
-    cell_record = fr.build_cell_feature_record(cell_features)
+    if cell_features["long_squares"]:
+        for s in cell_features["long_squares"]["subthreshold_sweeps"]:
```
Comment (NileGraddis) [inline ipfx/data_set_features.py]: This looks again like an already present issue, but we should not use single-letter variables.

File: ipfx/feature_record.py [hunk 7]
```diff
@@ -5,44 +5,46 @@
...
+        cell_record["vm_for_sag"] = nan_get(base, "vm_for_sag")
+        cell_record["f_i_curve_slope"] = nan_get(base, "fi_fit_slope")
+
+        # change the base to hero sweep
+        base = cell_features["long_squares"]["hero_sweep"]
+        cell_record["adaptation"] = nan_get(base, "adapt")
+        cell_record["latency"] = nan_get(base, "latency")
+        cell_record["avg_isi"] = nan_get(base, "mean_isi")
+
+        # mean feature of first spike for
```
Comment (NileGraddis) [inline ipfx/feature_record.py]: truncated?

File: ipfx/lab_notebook_reader.py [hunk 8]
```diff
@@ -67,7 +67,7 @@ def get_text_value(self, name, data_col, sweep_col, enable_col, sweep_num, defau
                     # this shouldn't happen, but if it does then bitch
                     #   as this situation hasn't been tested (eg, is
                     #   enabled indicated by 1.0, or "1.0" or "true" or ??)
-                    Exception("Enable flag not expected for text values")
+                    raise Exception("Enable flag not expected for text values")
```
Comment (NileGraddis) [inline ipfx/lab_notebook_reader.py]: This should be a [more specific kind of exception](https://docs.python.org/3/library/exceptions.html#concrete-exceptions).

File: ipfx/lims_queries.py [hunk 9]
```diff
@@ -201,3 +202,47 @@ def project_specimen_ids(project, passed_only=True):
...
+    res = query(sql)
+    if res:
+        err_dir = res[0]["storage_directory"]
+
+        file_list = glob.glob(os.path.join(err_dir, '*EPHYS_FEATURE_EXTRACTION_*_output.json'))
+        if file_list:
+            latest_file = max(file_list, key=os.path.getctime)   # get the most recent file
+            return latest_file
+        else:
+            return NO_OUTPUT_FILE
```
Comment (NileGraddis) [inline ipfx/lims_queries.py]: What is the rationale behind string error codes vs. e.g. exceptions here?

File: ipfx/lims_queries.py [hunk 10]
```diff
@@ -201,3 +202,47 @@ def project_specimen_ids(project, passed_only=True):
...
+    else:
+        return NO_SPECIMEN
+
+
+def fix_network_path(lims_path):
```
Comment (NileGraddis) [inline ipfx/lims_queries.py]: does this work correctly across platforms?

**Comment by tmchartrand** [comment]
@NileGraddis  I'll respond to a few specifics here but most of these were already existing issues, just code that moved between files maybe so it looks new? In general, I intended this as a hasty placeholder PR for what needs to happen rather than something polished.

File: ipfx/data_set_features.py (hunk 4 above)
Comment (tmchartrand) [reply ipfx/data_set_features.py]: yeah this is definitely out of scope for now, and I'm not sure whether it's actually worth the effort of fixing, I've only encountered in one specimen ever I think. I'll check in with ephys folks and create an issue if it seems worth fixing.

**Comment by NileGraddis** [comment]
@tmchartrand 

We greatly appreciate the work you've put in to fixing up this codebase. These comments are left so that we have a record of the work that needs to be done in order to merge this PR, not as a request for you specifically to do all this work yourself (though if there are some changes you would like to make, please feel free)!

**Comment by wbwakeman** [comment]
@tmchartrand This should be addressed by PR #449 which has been merged into the released version of ipfx.  Now available from pypi: `pip install --upgrade ipfx`

Please use that updated version and let us know if there are still issues that need to be addressed.

**Comment by tmchartrand** [comment]
@wbwakeman excited to try it! There were a couple minor fixes thrown in here that would likely still be useful, which I'll try to rebase and file at some point soon.

**Pull request closed without merging**